test/t8_forest/t8_gtest_forest_face_normal.cxx

```diff
@@ -0,0 +1,120 @@
+#include <t8_schemes/t8_default/t8_default_cxx.hxx>
+#include <t8_schemes/t8_default/t8_default_pyramid/t8_dpyramid_bits.h>
+#include <t8_cmesh/t8_cmesh_examples.h>
+#include <t8_forest/t8_forest_general.h>
+#include <t8_forest/t8_forest_geometrical.h>
+#include <test/t8_gtest_macros.hxx>
+
+/**
+ * This file tests the face normal computation of elements.
+ */
+
+class class_forest_face_normal: public testing::TestWithParam<std::tuple<t8_eclass_t, int>> {
+ protected:
+  void
+  SetUp () override
+  {
+    eclass = std::get<0> (GetParam ());
+    level = std::get<1> (GetParam ());
+    scheme = t8_scheme_new_default_cxx ();
+    t8_cmesh_t cmesh = t8_cmesh_new_hypercube (eclass, sc_MPI_COMM_WORLD, 0, 0, 0);
+    const int do_face_ghost = 1;
+    forest = t8_forest_new_uniform (cmesh, scheme, level, do_face_ghost, sc_MPI_COMM_WORLD);
+  }
+  void
+  TearDown () override
+  {
+    t8_forest_unref (&forest);
+  }
+  t8_forest_t forest;
+  t8_scheme_cxx *scheme;
+  t8_eclass_t eclass;
+  int level;
+};
+
+TEST_P (class_forest_face_normal, back_and_forth)
+{
+  /** Iterate over all elements of a uniformly refined forest. For all faceneighbors elements, if they are local,
+    * check if their facenormal is the negative of the corresponding facenormal of the neighbor elements.
+    */
+  const t8_locidx_t local_num_trees = t8_forest_get_num_local_trees (forest);
+  /* Iterate over all elements. */
+  for (t8_locidx_t itree = 0; itree < local_num_trees; itree++) {
+    const t8_locidx_t tree_elements = t8_forest_get_tree_num_elements (forest, itree);
+    const t8_eclass_t tree_eclass = t8_forest_get_tree_class (forest, itree);
+    ASSERT_EQ(eclass, tree_eclass);
+    const t8_eclass_scheme_c* escheme = t8_forest_get_eclass_scheme(forest, tree_eclass);
+    for (t8_locidx_t ielement = 0; ielement < tree_elements; ielement++) {
+      const t8_element_t *element = t8_forest_get_element_in_tree (forest, itree, ielement);
+      const int num_faces = escheme->t8_element_num_faces(element);
+      for (int iface = 0; iface < num_faces; iface++){
+        /* Compute facenormal */
+        double face_normal[3] = {0,0,0};
+        t8_forest_element_face_normal(forest,itree,element,iface,face_normal);
+
+        /* Get all face neighbors */
+
+        t8_element_t **neighbors;
+        int num_neighbors;
+        const int forest_is_balanced = 1;
+        t8_eclass_scheme_c *neigh_scheme;
+        int *dual_faces;
+        t8_locidx_t *neigh_ids;
+
+
+        t8_forest_leaf_face_neighbors (forest, itree, element, &neighbors, iface, &dual_faces, &num_neighbors,
+                                       &neigh_ids, &neigh_scheme, forest_is_balanced);
+
+        /* Iterate and compute their facenormal */
+
+        for(int ineigh = 0; ineigh < num_neighbors; ineigh++){
+          double neigh_face_normal[3];
+          t8_locidx_t ineightree;
+          /* Skip ghost elements, because those can't be passed to t8_forest_element_face_normal */
```

Comment:
Why can ghost elements not be passed to the face_normal function?

Can you not call `t8_forest_ghost_get_element` and call the face normal function on the returned element?

Reply:
ineightree currently needs to be a local tree for the t8_element_face_normal function

Reply:
That was not completely correct. Ineightree could apparently also be a ghost tree (i.e. `num_local_trees <= ghost_id < num_local_trees+num_ghost_trees`), but the `leaf_face_neighbors` function did not provide any information about the tree_id of the neighbor.

I extended that functionality to also return the global id of the neighbor tree if asked for, and introduced functionality to turn a global id into either local_tree or local_ghost id.

With this, the geometry functionality from t8_forest can also be made available for ghost elements.